File: vademecum/code-src/wzorce_projektowe_i_algorytmy/Dziedziczenie_polimorfizm_itp.cpp

// klasa abstrakcyjna - definiująca interfejs dla jakiejś grupy obiektów
// fakt bycia klasą abstrakcyjną wynika z niezdefiniowanej metody wirtualnej
// z faktu tego wynika niemożność utworzenia obiektów będących bezpośrednimi
// instancjami takiej klasy
class Ksztalt {
public:
	virtual float objetosc() = 0;
	// użycie virtual wymusza wywoływanie metody z klasy potomnej
	// kiedy odwołujemy się przy użyciu klasy bazowej
};

// klasy dziedzicząca po klasie Ksztalt
class Kula : public Ksztalt {
	float promien;
public:
	Kula(float r) {
		promien = r;
	}
	float objetosc() {
		return 1.25 * 3.14 * promien * promien * promien;
	}
};
class Szescian : public Ksztalt {
	float bok;
public:
	Szescian(float a) {
		bok = a;
	}
	float objetosc() {
		return bok*bok*bok;
	}
};

struct Kolor {
	char r, g, b;
};

class Cena {
public:
	Cena(float c);
};

class Material {
};

// klasa dziedzicząca po kilku klasach bazowych
struct Opakowanie : public Ksztalt, public Kolor, public Cena {
	Ksztalt  *ksztalt;
	Material *material;
public:
	Opakowanie(Ksztalt *k, const Kolor& kolor, float c) :
		// wywołanie konstruktorów klasy bazowej
		// jawnego dla Cena i kopiującego dla Kolor
		Cena(c), Kolor(kolor)
	{
		// inicjalizacja zmiennej wskaźnikowej ksztalt
		// dzięki rozwiązaniu z zmienną przechowującą informację o kształcie
		// nie ma potrzeby tworzenia różnych klas Opakowanie dla różnych kształtów
		ksztalt = k;
		material = 0;
	}

	Opakowanie(Ksztalt *k, float c) :
		// wywołanie konstruktora klasy bazowej Cena
		// oraz inicjalizacja zmiennych ksztalt i material
		Cena(c), ksztalt(k), material(0)
	{
		// inicjalizacja pól odziedziczonych po strukturze Kolor
		r = 0x96; g = 0xFF; b = 0x03;
	}

	float objetosc() {
		return ksztalt->objetosc();
	}

	// jest to alternatywna w stosunku co do dziedziczenia
	// metoda rozszerzania interfejsu jakiejś klasy
	
	// dziedziczenie (w przypadku dziedziczenia public) dodaje metody
	// klasy bazowej bezpośrednio do interfejsu klasy dziedziczącej
	// natomiast to rozwiązanie pozwala na pobranie poszczególnych klas
	// opisujących złożony obiekt niezależnie i operowaniu na ich interfejsie
	Material* getMaterial() {
		return material;
	}

	void setMaterial(Material *m) {
		material = m;
	}
};

File: vademecum/code-src/zmienne_i_struktury_danych/Dziedziczenie_polimorfizm_itp.py

# klasa abstracyjna
from abc import ABC, abstractmethod
class Kształt(ABC):
	@abstractmethod
	def objetosc(self):
		return None
	
	@staticmethod
	@abstractmethod
	def typKształtu():
		return None

# jednak ze względu na dynamiczne typowanie klasy abstrakcyjne
# w Pythonie odgrywają mniejszą rolę niż w C++

# klasy dziedzicząca po klasie Kształt
class Kula(Kształt):
	def __init__(self, r):
		self.promien = r
		
	def objetosc(self):
		return 1.25 * 3.14 * self.promien ** 3
	
	@staticmethod
	def typKształtu():
		return 'Kula'

class Prostopadłościan(Kształt):
	def __init__(self, a, b, c):
		self.a = a
		self.b = b
		self.c = c
	
	def objetosc(self):
		return self.a * self.b * self.c
	
	@staticmethod
	def typKształtu():
		return 'Prostopadłościan'

# można dziedziczyć po klasach nie abstrakcyjnych
# oraz po klasach, które już dziedziczą po jakiejś
class Szescian(Prostopadłościan):
	def __init__(self, a):
		self.bok = a;
	
	def objetosc(self):
		return self.bok ** 3

k = Kula(1)
print(k.typKształtu(), k.objetosc())

s = Szescian(2)
print(s.typKształtu(), s.objetosc())
# tu wykonała się funkcja `typKształtu` z klasy `Prostopadłościan`
# bo nie została nadpisana w klasie `Szescian`

class Kolor:
	def __init__(self, r, g, b):
		self.r = r
		self.g = g
		self.b = b

# Możliwe jest także dziedziczenie po jednej lub kilku klasach bazowych:
class Opakowanie(Szescian, Kolor):
	def __init__(self, a):
		Szescian.__init__(self, a)
		Kolor.__init__(self, .5, .5, .5)

o = Opakowanie(1)
print(o.objetosc(), o.b)

